driver-core/src/main/com/mongodb/internal/operation/QueryBatchCursor.java

```diff
@@ -320,83 +317,313 @@ private BsonDocument asGetMoreCommandDocument() {
-                                                                        connectionSource.getServerDescription().getAddress());
+    @Nullable
+    private ServerCursor initFromCommandResult(final BsonDocument getMoreCommandResultDocument) {
+        QueryResult<T> queryResult = getMoreCursorDocumentToQueryResult(getMoreCommandResultDocument.getDocument(CURSOR), serverAddress);
         postBatchResumeToken = getPostBatchResumeTokenFromResponse(getMoreCommandResultDocument);
         operationTime = getMoreCommandResultDocument.getTimestamp(OPERATION_TIME, null);
-        initFromQueryResult(queryResult);
+        return initFromQueryResult(queryResult);
     }
 
     private boolean limitReached() {
         return Math.abs(limit) != 0 && count >= Math.abs(limit);
     }
 
-    private void killCursor() {
-        if (serverCursor != null && killCursorOnClose) {
+    private BsonDocument getPostBatchResumeTokenFromResponse(final BsonDocument result) {
+        BsonDocument cursor = result.getDocument(CURSOR, null);
+        if (cursor != null) {
+            return cursor.getDocument(POST_BATCH_RESUME_TOKEN, null);
+        }
+        return null;
+    }
+
+    /**
+     * This class maintains all resources that must be released in {@link QueryBatchCursor#close()}.
+     * It also implements a {@linkplain #doClose() deferred close action} such that it is totally ordered with other operations of
+     * {@link QueryBatchCursor} (methods {@link #tryStartOperation()}/{@link #endOperation()} must be used properly to enforce the order)
+     * despite the method {@link QueryBatchCursor#close()} being called concurrently with those operations.
+     * This total order induces the happens-before order.
+     * <p>
+     * The deferred close action does not violate externally observable idempotence of {@link QueryBatchCursor#close()},
+     * because {@link QueryBatchCursor#close()} is allowed to release resources "eventually".
+     * <p>
+     * Only methods explicitly documented as thread-safe are thread-safe,
+     * others are not and rely on the total order mentioned above.
+     */
+    @ThreadSafe
+    private static final class ResourceManager {
+        private final MongoNamespace namespace;
+        private final Lock lock;
+        private volatile State state;
+        @Nullable
+        private volatile ConnectionSource connectionSource;
+        @Nullable
+        private volatile Connection pinnedConnection;
+        @Nullable
+        private volatile ServerCursor serverCursor;
+        private volatile boolean skipReleasingServerResourcesOnClose;
+
+        ResourceManager(final MongoNamespace namespace, @Nullable final ConnectionSource connectionSource,
+                @Nullable final Connection connectionToPin, @Nullable final ServerCursor serverCursor) {
+            this.namespace = namespace;
+            lock = new StampedLock().asWriteLock();
```

Comment:
For my edification: why is this preferable to ReentrantLock?

Reply:
Linking my previous [answer to this question](https://github.com/mongodb/mongo-java-driver/pull/722/files#r644208261). I am happy to discuss if needed.

Reply:
Nah, I just forgot.  Thanks!